### PR TITLE
Check that the secret key is non-zero modulo the order of the elliptic curve

### LINKED ISSUE #8783: BLS library does not properly check secret key Severity: Informational
### Description

The SecretKey.fromBytes method performs some validation on the input, such as ensuring the secret key is non-zero and of correct length. However, the check against the secret key is not done modulo the order of the elliptic curve. Therefore, it is possible for a secret key equivalent to zero to be used if the value of the secret key is a multiple of this elliptic curve order.

https://github.com/ChainSafe/blst-ts/blob/ad0af59d9d01d45ea144b8306f44e45426c5888b/src/lib.ts#L63-L73

### Acceptance Criteria

Add checks to all functions making use of the SecretKey.fromBytes function to ensure that the secret key being used for signing and other operations is non-zero modulo the order of the elliptic curve.

### Additional Information
Reference: https://github.com/ChainSafe/bls/issues/96

### PR DESCRIPTION
**Pull request by mosmartin**
### What was the problem?

This PR resolves #8783

### How was it solved?

Added the `isSecretKeyNonZeroModEC` function to check that the secret key generated using the `SecretKey.fromBytes` function is non-zero modulo the order of the elliptic curve.

### How was it tested?

Added unit tests

### CODE REVIEW

File: elements/lisk-cryptography/src/bls_lib/lib.ts [hunk 1]
```diff
@@ -45,9 +45,31 @@ export const blsKeyValidate = (pk: Buffer): boolean => {
 // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-2.3
 export const blsKeyGen = (ikm: Buffer): Buffer => Buffer.from(SecretKey.fromKeygen(ikm).toBytes());
 
+// check that the secret key generated is non-zero modulo the order of the elliptic curve.
+export const isSecretKeyNonZeroModEC = (secretKey: SecretKey): boolean => {
+	// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-10#section-4.2.1
+	const curveOrder = '0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001';
```
Comment (shuse2) [inline elements/lisk-cryptography/src/bls_lib/lib.ts:51]: I couldn't tell why this value is used for checking modulo from the spec

**Review by shuse2 — APPROVED** [review]
Other than correctness of the value used, LGTM

File: elements/lisk-cryptography/src/bls_lib/lib.ts (hunk 1 above)
Comment (mosmartin) [reply elements/lisk-cryptography/src/bls_lib/lib.ts:51]: Oh maybe I should point the link to the exact value. the [order](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-10#section-4.2.1-11.4) is `r = 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001`. Thanks to @AndreasKendziorra who pointed this out to me.

File: elements/lisk-cryptography/test/bls_lib/lib.spec.ts [hunk 2]
```diff
@@ -58,6 +60,29 @@ interface EthFastAggrVerifySpec {
...
 describe('bls_lib', () => {
+	const curveOrder = '0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001';
+
+	describe('isSecretKeyNonZeroModEC', () => {
+		it('should return true when given a valid secret key', () => {
+			const secretKey = SecretKey.fromBytes(Buffer.alloc(32, 1));
+			expect(isSecretKeyNonZeroModEC(secretKey)).toBe(true);
+		});
+
+		it('should return true when given a non-zero modulo secret key', () => {
```
Comment (shuse2) [inline elements/lisk-cryptography/test/bls_lib/lib.spec.ts:71]: I think we should add test case for zero secret key?

File: elements/lisk-cryptography/test/bls_lib/lib.spec.ts [hunk 3]
```diff
@@ -68,24 +93,76 @@ describe('bls_lib', () => {
...
+
+		it('should throw an error if the input buffer is not 32 bytes long', () => {
+			const sk = Buffer.alloc(31, 1);
+			expect(() => blsSkToPk(sk)).toThrow(ErrorBLST);
+		});
 	});
 
 	describe('blsSign', () => {
-		// Signing with the zero private key is not a use case according to the BLS specifications
 		describe.each(getAllFiles(['eth2_bls_specs/sign', 'bls_specs/sign'], /sign_case_zero_privkey/))(
```
[the remark this answers is not in the document]
Comment (AndreasKendziorra) [reply elements/lisk-cryptography/test/bls_lib/lib.spec.ts:123]: here the test from Ethereum for the zero secret key was manually excluded, which should not be done anymore.

File: elements/lisk-cryptography/test/bls_lib/lib.spec.ts [hunk 4]
```diff
@@ -68,24 +93,76 @@ describe('bls_lib', () => {
...
+					it('should throw an error if the private key is all zeros', () => {
+						expect(() => blsSign(hexToBuffer(privkey), hexToBuffer(message))).toThrow(
+							'ZERO_SECRET_KEY',
+						);
+					});
+				}
 			},
 		);
+
+		it('should throw an error when given an input buffer with value equal to the curve order', () => {
```
Comment (AndreasKendziorra) [inline elements/lisk-cryptography/test/bls_lib/lib.spec.ts:146]:
```suggestion
		it('should throw an error when given an input buffer with value equal to the group order', () => {
```

File: elements/lisk-cryptography/test/bls_lib/lib.spec.ts [hunk 5]
```diff
@@ -68,24 +93,76 @@ describe('bls_lib', () => {
...
+			expect(() => blsSign(sk, message)).toThrow('Secret key is not valid.');
+		});
+
+		it('should throw an error when a secret key that is non-zero but zero modulo the group order', () => {
+			// sk equals 2*r where r is order of the groups G1 and G2
+			const sk = Buffer.from(
+				'e7db4ea6533afa906673b0101343b00aa77b4805fffcb7fdfffffffe00000002',
+				'hex',
+			);
+			const message = Buffer.from(
```
Comment (AndreasKendziorra) [inline elements/lisk-cryptography/test/bls_lib/lib.spec.ts:159]: feel free to use any other message if you prefer, e.g. like in line 148

File: elements/lisk-cryptography/test/bls_lib/lib.spec.ts [hunk 6]
```diff
@@ -181,6 +258,14 @@ describe('bls_lib', () => {
 				);
 			});
 		});
+
+		it('should throw an error when given an input buffer with value equal to the curve order', () => {
```
Comment (AndreasKendziorra) [inline elements/lisk-cryptography/test/bls_lib/lib.spec.ts:262]:
```suggestion
		it('should throw an error when given an input buffer with value equal to the group order', () => {
```

File: elements/lisk-cryptography/test/bls_lib/lib.spec.ts (hunk 2 above)
Comment (mosmartin) [reply elements/lisk-cryptography/test/bls_lib/lib.spec.ts:71]: That would be a nice check but I think the functionality encapsulated in this function is only to check that a secret key being used for signing and other operations is non-zero modulo the order of the group. Checks for zero secret key are carried out in conjunction with `isSecretKeyNonZeroModEC` function. Thoughts?

File: elements/lisk-cryptography/src/bls_lib/lib.ts [hunk 7]
```diff
@@ -45,9 +45,31 @@ export const blsKeyValidate = (pk: Buffer): boolean => {
 // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-2.3
 export const blsKeyGen = (ikm: Buffer): Buffer => Buffer.from(SecretKey.fromKeygen(ikm).toBytes());
 
+// check that the secret key generated is non-zero modulo the order of the elliptic curve.
```
[the remark this answers is not in the document]
Comment (mosmartin) [reply elements/lisk-cryptography/src/bls_lib/lib.ts:48]: Ahh yes. Let me update this.

File: elements/lisk-cryptography/src/bls_lib/lib.ts [hunk 8]
```diff
@@ -123,8 +145,14 @@ export const blsPopProve = (sk: Buffer): Buffer => {
 	const message = blsSkToPk(sk);
 	const sig = new blst.P2();
 
+	const secretKey = SecretKey.fromBytes(sk);
+
+	if (!isSecretKeyNonZeroModEC(secretKey)) {
+		throw new Error('Secret key is not valid.');
+	}
```
[the remark this answers is not in the document]
Comment (mosmartin) [reply elements/lisk-cryptography/src/bls_lib/lib.ts:152]: If we remove this check from here, `SecretKey.fromBytes(sk)` on L148 is not checked for compliance that it should be non-zero modulo the group order.

[the remark this answers is not in the document]
Comment (AndreasKendziorra) [reply elements/lisk-cryptography/src/bls_lib/lib.ts]: I'm not sure I understand. In `blsSkToPk` you do already
```js
        const secretKey = SecretKey.fromBytes(sk);

	if (!isSecretKeyNonZeroModOrder(secretKey)) {
		throw new Error('Secret key is not valid.');
	}
```
and `blsSkToPk` is called in line 143. So, I'm sure you could remove lines 148-150 and all tests for `blsPopProve` should pass including the test for`sk = groupOrder`.

BTW: It would be good to add the test for `blsPopProve` with `sk=0`.

File: elements/lisk-cryptography/test/bls_lib/lib.spec.ts (hunk 2 above)
Comment (AndreasKendziorra) [reply elements/lisk-cryptography/test/bls_lib/lib.spec.ts]: can be resolved from my side

File: elements/lisk-cryptography/src/bls_lib/lib.ts (hunk 1 above)
Comment (AndreasKendziorra) [reply elements/lisk-cryptography/src/bls_lib/lib.ts]: can be resolved from my side

File: elements/lisk-cryptography/test/bls_lib/lib.spec.ts [hunk 9]
```diff
@@ -58,140 +60,223 @@ interface EthFastAggrVerifySpec {
...
+				});
 			});
 		});
-	});
 
-	describe('blsPopProve', () => {
-		describe.each(getAllFiles(['bls_specs/pop_prove']))('%s', ({ path }) => {
-			it('should create valid proof of possession', () => {
-				const { input, output } = loadSpecFile<{ input: string; output: string }>(path);
+		describe('blsPopProve', () => {
```
Comment (AndreasKendziorra) [inline elements/lisk-cryptography/test/bls_lib/lib.spec.ts:250]: As mentioned above, at test for `sk = 0` would be good.

**Review by AndreasKendziorra — CHANGES_REQUESTED** [review]

**Review by AndreasKendziorra — APPROVED** [review]

File: elements/lisk-cryptography/src/bls_lib/lib.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mosmartin) [reply elements/lisk-cryptography/src/bls_lib/lib.ts]: `sk` is only passed as an arg to `blsSkToPk` at L143. if we remove the lines 148-150, then the `sk` at L146 is not checked or it's safe to assume that after L143, we do not need to check if `SecretKey.fromBytes(sk);` is "safe"?
